src/fury/meta/preprocessor.h

```diff
@@ -59,68 +70,68 @@
-#define FURY_PP_FOREACH_IMPL_61(X, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61) X(_1) X(_2) X(_3) X(_4) X(_5) X(_6) X(_7) X(_8) X(_9) X(_10) X(_11) X(_12) X(_13) X(_14) X(_15) X(_16) X(_17) X(_18) X(_19) X(_20) X(_21) X(_22) X(_23) X(_24) X(_25) X(_26) X(_27) X(_28) X(_29) X(_30) X(_31) X(_32) X(_33) X(_34) X(_35) X(_36) X(_37) X(_38) X(_39) X(_40) X(_41) X(_42) X(_43) X(_44) X(_45) X(_46) X(_47) X(_48) X(_49) X(_50) X(_51) X(_52) X(_53) X(_54) X(_55) X(_56) X(_57) X(_58) X(_59) X(_60) X(_61)
-#define FURY_PP_FOREACH_IMPL_62(X, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62) X(_1) X(_2) X(_3) X(_4) X(_5) X(_6) X(_7) X(_8) X(_9) X(_10) X(_11) X(_12) X(_13) X(_14) X(_15) X(_16) X(_17) X(_18) X(_19) X(_20) X(_21) X(_22) X(_23) X(_24) X(_25) X(_26) X(_27) X(_28) X(_29) X(_30) X(_31) X(_32) X(_33) X(_34) X(_35) X(_36) X(_37) X(_38) X(_39) X(_40) X(_41) X(_42) X(_43) X(_44) X(_45) X(_46) X(_47) X(_48) X(_49) X(_50) X(_51) X(_52) X(_53) X(_54) X(_55) X(_56) X(_57) X(_58) X(_59) X(_60) X(_61) X(_62)
-#define FURY_PP_FOREACH_IMPL_63(X, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62, _63) X(_1) X(_2) X(_3) X(_4) X(_5) X(_6) X(_7) X(_8) X(_9) X(_10) X(_11) X(_12) X(_13) X(_14) X(_15) X(_16) X(_17) X(_18) X(_19) X(_20) X(_21) X(_22) X(_23) X(_24) X(_25) X(_26) X(_27) X(_28) X(_29) X(_30) X(_31) X(_32) X(_33) X(_34) X(_35) X(_36) X(_37) X(_38) X(_39) X(_40) X(_41) X(_42) X(_43) X(_44) X(_45) X(_46) X(_47) X(_48) X(_49) X(_50) X(_51) X(_52) X(_53) X(_54) X(_55) X(_56) X(_57) X(_58) X(_59) X(_60) X(_61) X(_62) X(_63)
+#define FURY_PP_FOREACH_IMPL_2(X, _1, _2) X(_1), X(_2)
+#define FURY_PP_FOREACH_IMPL_3(X, _1, _2, _3) X(_1), X(_2), X(_3)
+#define FURY_PP_FOREACH_IMPL_4(X, _1, _2, _3, _4) X(_1), X(_2), X(_3), X(_4)
+#define FURY_PP_FOREACH_IMPL_5(X, _1, _2, _3, _4, _5) X(_1), X(_2), X(_3), X(_4), X(_5)
+#define FURY_PP_FOREACH_IMPL_6(X, _1, _2, _3, _4, _5, _6) X(_1), X(_2), X(_3), X(_4), X(_5), X(_6)
+#define FURY_PP_FOREACH_IMPL_7(X, _1, _2, _3, _4, _5, _6, _7) X(_1), X(_2), X(_3), X(_4), X(_5), X(_6), X(_7)
```

Comment:
Is it necessary to change? This modification will reduce the generalization of this macro.

Reply:
Compilation fails on C++ without modification, syntax error.

Reply:
If there are some compile errors, could you figure out the root cause and search for a better solution?

You can refer to the implementation of [Boost.Preprocessor](https://www.boost.org/doc/libs/1_84_0/libs/preprocessor/doc/index.html).

Also, could you tell me the MSVC version you are using?

Reply:
VS2022(14.38.33130)

Reply:
```
src/fury/meta/preprocessor_test.cc(48): error C2146: 语法错误: 缺少“;”(在标识符“x”的前面)
src/fury/meta/preprocessor_test.cc(49): error C2146: 语法错误: 缺少“;”(在标识符“x”的前面)
src/fury/meta/preprocessor_test.cc(50): error C2146: 语法错误: 缺少“;”(在标识符“x”的前面)
src/fury/meta/preprocessor_test.cc(51): error C2146: 语法错误: 缺少“;”(在标识符“x”的前面)
src/fury/meta/preprocessor_test.cc(57): error C2607: 静态断言失败
src/fury/meta/preprocessor_test.cc(58): error C2131: 表达式的计算结果不是常数
src/fury/meta/preprocessor_test.cc(58): note: 超出范围的索引 1 导致失败；允许范围是 0 <= 索引 < 1
src/fury/meta/preprocessor_test.cc(59): error C2131: 表达式的计算结果不是常数
src/fury/meta/preprocessor_test.cc(59): note: 超出范围的索引 2 导致失败；允许范围是 0 <= 索引 < 1
```

This error occurs when calling FURY_PP_FOREACH.

Reply:
> It will reduce the generalization of this macro.

I don’t quite understand the problem you mentioned. The following code is also like this.

```
#define FURY_PP_NARG_REV()                                                     \
  63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 47, 46, 45,  \
      44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27,  \
      26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9,   \
      8, 7, 6, 5, 4, 3, 2, 1, 0
```

Reply:
Previously it does not need to be seperated by comma (`,`).

Reply:
So there is no problem with this modification here.

Reply:
I cannot see any connection between it and `FURY_PP_NARG_REV`.

Reply:
> So there is no problem with this modification here.

Previously, this macro could be expanded in other ways, such as into symbols connected by a plus sign. Adding a comma makes it impossible. 

I don't agree with this change. You'll need to look at Boost.Preprocessor to see more appropriate changes.